lib/sycamore/sycamore/transforms/llm_query.py

```diff
@@ -0,0 +1,94 @@
+from abc import ABC, abstractmethod
+from typing import Optional
+
+from sycamore.data import Element, Document
+from sycamore.llms.openai import OpenAIModels
+from sycamore.plan_nodes import NonCPUUser, NonGPUUser, Node
+from sycamore.llms import OpenAI, OpenAIClientWrapper
+from sycamore.transforms.map import Map
+from sycamore.utils.time_trace import timetrace
+
+
+class LLMQueryAgent(ABC):
+    @abstractmethod
+    def execute_query(self, document: Document) -> Document:
+        pass
+
+
+class LLMTextQueryAgent(LLMQueryAgent):
+    """
+    LLMTextQueryAgent uses a specified LLM to execute LLM queries about a document
+
+    Args:
+        prompt: A prompt to be passed into the underlying LLM execution engine
+        openai_model: (Optional) The type of OpenAI model to be used in the execution. Defaults to GPT-4O
+        client_wrapper: (Optional) Specifications for the OpenAI client wrapper. Ignored if openai_model is specified
+        output_property: (Optional) The output property to add results in. Defaults to 'llm_response'
+        llm_kwargs: (Optional) LLM keyword argument for the underlying execution engine
+        per_element: (Optional) Whether to execute the call per each element or on the Document itself. Defaults to
+        True.
+
+    Example:
+         .. code-block:: python
+
+            prompt="Tell me the important numbers from this element"
+            llm_query_agent = LLMElementTextSummarizer(prompt=prompt)
+
+            context = sycamore.init()
+            pdf_docset = context.read.binary(paths, binary_format="pdf")
+                .partition(partitioner=UnstructuredPdfPartitioner())
+                .llm_query(query_agent=llm_query_agent)
+    """
+
+    model = OpenAIModels.GPT_4O
+
+    def __init__(
+        self,
+        prompt: str,
```

Comment:
One common pattern is to have a prompt that is parameterized based on the element text or properties. Consider a prompt like [this](https://github.com/aryn-ai/pimco-scripts/blob/b8e206f9560d1538c29af5e5b9b7b7c8a5901193/pimco_scripts/questions.py#L349), which is parameterized not only by the text, but also the `company`, `quarter`, and `year`. We should think about how to support that here. Maybe pass in the properties dict + text representation to str.format? An alternative would be to use something like Jinja (which @HenryL27 used [here](https://github.com/aryn-ai/sycamore/blob/de50a997047aeccf4acfd88aa5cefd27192ea243/lib/sycamore/sycamore/transforms/augment_text.py#L54)). I think a good goal would be to support all of the llm calls in that first repo without custom llm code.